### DISCUSSION TITLE
WithViewStore not updating state and view, even if Store is updated

### DISCUSSION THREAD

**Discussion by drucelweisse** [pointfreeco/swift-composable-architecture#1628] [Q&A]
Hello, recently I started adopting TCA and tried to refactor simple view from mvvm that fetches texts from file url. My problem is that `WithViewStore` doesn't update on `Store` changes. Maybe I miss something or do not understand?

Here's an example

```swift
public struct DetailInfo: ReducerProtocol {
    
    @Dependency(\.locale) var locale
    @Dependency(\.fileClient) var fileClient

    public struct State: Equatable {
        var texts: [String]

        public init(texts: [String] = []) {
            self.texts = texts
        }
        
        public static let initial = State()
    }

    public enum Action: Equatable {
        case fetch
        case loadedTexts([String])
    }

    public func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
        switch action {
        case .fetch:
            return .task {
                let url = Resources.exampleURL
                let dict = try await fileClient.loadDecodable(type: [String: [String]].self, from: url)
                let texts = dict[locale.languageCode!]!
                return .loadedTexts(texts)
            }
        case let .loadedTexts(texts):
            state.texts = texts
            return .none
        }
    }
    
    public init() {
        
    }
}
```

View body

```swift
public var body: some View {
        WithViewStore(store) { viewStore in
            ScrollView(showsIndicators: false) {
                VStack(spacing: metrics.contentStackSpacing) {
                    image
                    VStack(alignment: .leading, spacing: metrics.headerToContentSpacing) {
                        header
                        content(texts: viewStore.texts)
                    }
                    .padding(.horizontal)
                    
                }
                .padding(.top, metrics.containerTopPadding)
                .padding(.bottom, metrics.containerBottomPadding)
            }
            .background(style.backgroundColor.ignoresSafeArea())
            .task {
                await viewStore.send(.fetch).finish()
            }
        }
}
```

And Logs:
```swift
WithViewStore< DetailInfo.State, DetailInfo.Action, _>@DetailInfo Feature/DetailInfoView.swift:31 (Initial state)
  DetailInfo.State(texts: [])
WithViewStore< DetailInfo.State, DetailInfo.Action, _>@DetailInfoFeature/DetailtInfoView.swift:31 (Initial state)
  DetailtInfo.State(texts: [])
received action:
  DetailInfo.Action.fetch
  (No state changes)
received action:
  DetailInfo.Action.loadedTexts(
    [
      [0]: "Text",
    ]
  )
- DetailInfo.State(texts: [])
+ DetailInfo.State(
+   texts: [
+     [0]: "Text",
+   ]
+ )
```

**Comment by lukeredpath** [comment]
How does this reducer fit in with the rest of your reducer structure and how are you creating the store that you pass into this view?

**Reply by drucelweisse** [reply]
This is the first reducer of application so it's really simple. It's inside closure for sheet content

`DetailInfoView(store: .init(initialState: .initial, reducer: DetailInfo()))`

**Reply by tgrapperon** [reply]
If this `DetailInfoView` is recreated for some reason, you will restart with a new `.initial` state and get the impression that nothing is updated. I suppose that this is what @lukeredpath is considering. I would make sure that you hold a reference to this `Store` instance somewhere. I suppose that the correct place to hold the `Store` instance is in the presenting `View` (or its model) as you present it in a sheet, but you can make a quick check by storing it in a top level variable to see if it fixes your updating issue.

**Reply by drucelweisse** [reply]
Yes, thank you all. I forgot that store is a reference type. Added to presenting View.

**Reply by tgrapperon** [reply]
@drucelweisse Did this fix your updating issue?

**Reply by drucelweisse** [reply]
I'm not sure about view recreation, `Self._printChanges()` didn't show anything, but holding a reference to Store fixed that problem

**Reply by mbrandonw** [reply]
@drucelweisse just to reiterate what @tgrapperon mentioned, you should aim to have a single `Store` at the root of your application, and "scopes" of the store should be passed down to child views by using the [`scope`][scope-docs] method on `Store`.

Further, the initializer `WithViewStore(store)` is discouraged due to the fact that it observes _all_ state in the store. Instead, you should use the [`WithViewStore.init(_:observe:)`][withviewstore-init-docs] initializer, which requires you to explicitly specify what state you want to observe. 

You can also check out our dedicated article on [performance][performance-article] to read more about why we recommend this.

[scope-docs]: https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/store/scope(state:action:)
[withviewstore-init-docs]: https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/withviewstore/init(_:observe:content:file:line:)
[performance-article]: https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/performance#View-stores

**Reply by tgrapperon** [reply]
@drucelweisse Indeed, I assumed that you were partially converting a vanilla app to TCA, and that this `Detail` feature was the first entry point of TCA in your app, but if you're powering your whole app with TCA (which is the preferred approach), you should definitely scope stores as you drill down into your app's domains, as @mbrandonw said.

**Reply by drucelweisse** [reply]
Yeah, it's the first entry of TCA in the vanilla SwiftUI app, I'm going to make transition step by step. Thanks again! Can we close this discussion?
